Tables: Table log modes (#1335)
wandb/docs: content/guides/models/tables/log_tables.md

You [create and log tables]({{< relref "#create-and-log-a-new-table" >}}) using the W&B Python SDK. You can make a table [immutable, mutable, or incremental]({{< relref "#table-logging-modes" >}}). The mode you set determines how if and how you can modify the table after it is created.

## Create a table

Log a table to W&B using the `wandb.Table` class. 

1. First, create a new run with `wandb.init()`. 
2. Next, create a table object with `wandb.Table`. Specify the columns and data for the table for the `columns` and `data` parameters, respectively. 
3. Finally, log the table to W&B with `run.log()`.

The following example shows how to create and log a table with two columns, `a` and `b`, and two rows of data, `["a1", "b1"]` and `["a2", "b2"]`:

```python
import wandb

with wandb.init(project="table-test") as run:
    my_table = wandb.Table(
        columns=["a", "b"], data=[["a1", "b1"], ["a2", "b2"]],
        log_mode="IMMUTABLE"
    )
    run.log({"Table Name": my_table})
```

## Logging modes

> There are three table modes. The mode you set when you create a table determines how if and how you can modify the table after it is created. The three modes are:

Reviewer: Added a suggestion, not crazy about it, but I think we need to capture that it:
1. Changes how the table python object can be modified
2. Changes how it is logged to W&B

```suggestion
There are three table modes. The mode you set when you create a table determines how you can modify the table after it is created and if it can be logged again. The three modes are:
```

- `IMMUTABLE` mode (default): Once a table is logged, it cannot be modified.
- `MUTABLE` mode: Update an existing table by replacing the existing table with a new one.

> - `INCREMENTAL` mode: Add batches of new rows to a table iteratively.

Reviewer: I think there might be a lack of understanding on my part, but when I think about the Table, I'm looking it through the lens of a python object. That python object can then be logged to W&B.

Maybe the disconnect is that our docs generally talk about WBMedia being created in the W&B system right and not just python object creation?

I'd still like to communicate somehow that:
1. IMMUTABLE tables can't be logged multiple times
2. MUTABLE tables can be logged multiple times, each time creating a new artifact version of the table. I don't think I'd call it an update
3. INCREMENTAL - each time the table is logged, a new table increment is created basically a smaller table

{{% alert %}}
The following example uses a placeholder function `load_eval_data()` to load data and a placeholder function `model.predict()` to make predictions. You will need to replace these with your own data loading and prediction functions.
{{% /alert %}}

```python
import wandb

run = wandb.init(project="mutable-table-demo")

# Create a table object with MUTABLE logging mode
table = wandb.Table(columns=["input", "label", "prediction"],
                    log_mode="MUTABLE")

# Load data and make predictions
inputs, labels = load_eval_data() # Placeholder function
raw_preds = model.predict(inputs) # Placeholder function

for inp, label, pred in zip(inputs, labels, raw_preds):
    table.add_data(inp, label, pred)

# Step 1: Log initial data 
wandb.log({"eval_table": table})  # Log initial table

# Step 2: Add confidence scores
confidences = compute_confidences(raw_preds)
table.add_column("confidence", confidences)
wandb.log({"eval_table": table})  # Updates the table with new column

# Step 3: Add post-processed predictions
post_preds = postprocess_predictions(raw_preds, confidences)

> table.add_column("final_prediction", post_preds)

Reviewer: Left a suggestion below on how we can improve this example

{{% alert %}}

The following example uses a placeholder function `get_training_batch()` to load data, a placeholder function `train_model_on_batch()` to train the model, and a placeholder function `predict_on_batch()` to make predictions. You will need to replace these with your own data loading, training, and prediction functions.

{{% /alert %}}

```python
import wandb

run = wandb.init(project="incremental-table-demo")

# Create a table with INCREMENTAL logging mode
table = wandb.Table(columns=["step", "input", "label", "prediction"],
                    log_mode="INCREMENTAL")

# Training loop
for step in range(get_num_batches()): # Placeholder function
    # Load batch data
    inputs, labels = get_training_batch(step) # Placeholder function

    # Train and predict
    train_model_on_batch(inputs, labels) # Placeholder function
    predictions = predict_on_batch(inputs) # Placeholder function

    # Add batch data to table
    for input_item, label, prediction in zip(inputs, labels, predictions):
        table.add_data(step, input_item, label, prediction)

    # Log the table incrementally
    wandb.log({"training_table": table}, step=step)

run.finish()
```

> Each call to `wandb.log()` appends new rows to an existing table. In the previous code example, the table is logged once per training step. The `step` parameter is used to specify the training step for each log call. The table is updated with new rows for each batch of data processed in the training loop. 

Reviewer: I think we can change "Each call to `wandb.log()` appends new rows to an existing table" into something that communicates that we're taking that increment / chunk of data and creating a new artifact entry for it.

So typically, tables are represent as a singular JSON file. For incremental tables that were logged N times, we'll create N json files.

As an example, consider the follow code snippet that combines `INCREMENTAL` and `IMMUTABLE` logging modes:

```python
import wandb

run = wandb.init(project="combined-logging-example")

# Create an incremental table for efficient updates during training
incr_table = wandb.Table(columns=["step", "input", "prediction", "label"],
                         log_mode="INCREMENTAL")

# Training loop
for step in range(get_num_batches()):
    # Process batch
    inputs, labels = get_training_batch(step)
    predictions = model.predict(inputs)

    # Add data to incremental table
    for inp, pred, label in zip(inputs, predictions, labels):
        incr_table.add_data(step, inp, pred, label)

    # Log the incremental update (suffix with -incr to distinguish from final table)
    run.log({"table-incr": incr_table}, step=step)

# At the end of training, create a complete immutable table with all data
# Using the default IMMUTABLE mode to preserve the complete dataset
final_table = wandb.Table(columns=incr_table.columns, data=incr_table.data, log_mode="IMMUTABLE")
run.log({"table": final_table})

run.finish()
```

> In the previous example, the `incr_table` is logged incrementally (with `log_mode="INCREMENTAL"`) during training. Each time a new batch is processed, the table is updated with new rows. This allows you to monitor the training process in real-time and visualize the incremental updates in the W&B App.

Reviewer: I think we should remove the "real-time" part because I don't know what real-time means to our users. I don't think it's instantaneous.

Maybe we can just say it allows you to view increments as they come in during training?

> At the end of training, a complete immutable table is created with all data from the incremental table. This allows you to preserve the complete dataset while still benefiting from the efficiency of incremental logging during training.

Reviewer: The technical motivation for using both incremental and immutable logging modes is because stitching together 100 JSON files on the backend can potentially take awhile, so manipulating that table and performing analysis can be rough during the initial uncached load.

So ideally, use incrementally logged tables for viewing updates and then log the immutable table at the very end for analysis

## Examples 

### Enriching evaluation results with MUTABLE

```python
run = wandb.init(project="mutable-logging")

# Step 1: Log initial predictions
table = wandb.Table(columns=["input", "label", "prediction"], log_mode="MUTABLE")
inputs, labels = load_eval_data()
raw_preds = model.predict(inputs)

for inp, label, pred in zip(inputs, labels, raw_preds):
    table.add_data(inp, label, pred)

run.log({"eval_table": table})  # Log raw predictions

> # Step 2: Add confidence scores (e.g. max softmax)
> confidences = compute_confidences(raw_preds)
> table.add_column("confidence", confidences)
> run.log({"eval_table": table})  # Add confidence info
> 
> # Step 3: Add post-processed predictions
> # (e.g., thresholded or smoothed outputs)
> post_preds = postprocess_predictions(raw_preds, confidences)
> table.add_column("final_prediction", post_preds)

Reviewer: I think this example can be improved by using a more concrete example for computing confidences and post processing predictions

```suggestion
# Step 2: Add confidence scores (e.g. max softmax)
confidences = np.max(raw_preds, axis=1)
table.add_column("confidence", confidences)
run.log({"eval_table": table})  # Add confidence info

# Step 3: Add post-processed predictions
# (e.g., thresholded or smoothed outputs)
post_preds = (confidences > 0.7).astype(int)
table.add_column("final_prediction", post_preds)
```

- `MUTABLE`: After you log the table to W&B, you can overwrite the existing table by replacing that table with a new one.
- `INCREMENTAL`: Add batches of new rows to a table throughout the machine learning experiment.

The following table describes the differences between the three modes and common use cases for each:

| Mode  | Use Cases  | Benefits  |
| ----- | ---------- | ----------|
| `IMMUTABLE`   | - Storing tabular data generated at the end of a run for further analysis                              | - Minimal overhead when logged at the end of a run<br>- All rows rendered in UI |
| `MUTABLE`     | - Adding columns to existing tables<br>- Enriching results with new information                        | - Capture Table mutations<br>- All rows rendered in UI                          |
| `INCREMENTAL` | - Long-running training jobs<br>- Processing large datasets in batches<br>- Monitoring ongoing results | - View updates on UI during training<br>- Ability to step through increments    |

The next sections show example code snippets for each mode along with considerations for when to use each mode.

### MUTABLE mode

> `MUTABLE` mode updates an existing table by replacing the existing table with a new one. `MUTABLE` mode is useful when you want to add new columns and rows to an existing table in a non iterative process. If you want to add new rows incrementally like in a training loop, consider using [`INCREMENTAL` mode]({{< relref "#INCREMENTAL-mode" >}}) instead.

Reviewer: ```suggestion
`MUTABLE` mode updates an existing table by replacing the existing table with a new one. `MUTABLE` mode is useful when you want to add new columns and rows to an existing table in a non iterative process. If you want to add new batches of rows incrementally like in a training loop, consider using [`INCREMENTAL` mode]({{< relref "#INCREMENTAL-mode" >}}) instead.
```

```python
import wandb
import numpy as np

run = wandb.init(project="mutable-table-demo")

# Create a table object with MUTABLE logging mode
table = wandb.Table(columns=["input", "label", "prediction"],
                    log_mode="MUTABLE")

# Load data and make predictions
inputs, labels = load_eval_data() # Placeholder function
raw_preds = model.predict(inputs) # Placeholder function

for inp, label, pred in zip(inputs, labels, raw_preds):
    table.add_data(inp, label, pred)

# Step 1: Log initial data 
wandb.log({"eval_table": table})  # Log initial table

# Step 2: Add confidence scores (e.g. max softmax)
confidences = np.max(raw_preds, axis=1)
table.add_column("confidence", confidences)
run.log({"eval_table": table})  # Add confidence info

# Step 3: Add post-processed predictions
# (e.g., thresholded or smoothed outputs)
post_preds = (confidences > 0.7).astype(int)
table.add_column("final_prediction", post_preds)
wandb.log({"eval_table": table})  # Final update with another column

run.finish()
```
In the previous example, the table is logged three times: once with the initial data, once with the confidence scores, and once with the final predictions.

{{% alert %}}

> Internally, the table is replaced each time you log the table. Overwriting a table with a new one is computationally expensive and can be slow for large tables. If you are logging a large number of rows, consider using `INCREMENTAL` mode instead.

Reviewer: ```suggestion
Internally, the table is replaced each time you log the table. Overwriting a table with a new one is computationally expensive and can be slow for large tables. If you are logging a large number of rows, consider logging the final table at the very end and / or using `INCREMENTAL` mode to log batches of rows.
```

> Internally, the table is replaced each time you log the table. Overwriting a table with a new one is computationally expensive and can be slow for large tables. If you are logging a large number of rows, consider using `INCREMENTAL` mode instead.

Reviewer: On second thought, I think we should remove this alert and place it somewhere else. Because this example deals with adding columns to a table. Users can't add columns to a table with log_mode= incremental

Author: Good callout. Removed alert. 

Double checking: 
* with mutable you can add columns or rows
* incremental you can only add rows

Reviewer: for incremental tables, only new rows of data can be added

Author: Oops. Sorry. Meant to write rows 🤦

Reviewer: 😄 I figured it was a typo!!

{{% /alert %}}

### INCREMENTAL mode

In incremental mode, you log batches of rows to a table throughout the machine learning experiment. This

> is ideal for monitoring long-running jobs or when working with large datasets that would be inefficient to log all at once.

Reviewer: ```suggestion
is ideal for monitoring long-running jobs or when working with large tables that would be inefficient to log during the run for updates.
```

{{% /alert %}}

```python
import wandb

run = wandb.init(project="incremental-table-demo")

# Create a table with INCREMENTAL logging mode
table = wandb.Table(columns=["step", "input", "label", "prediction"],
                    log_mode="INCREMENTAL")

# Training loop
for step in range(get_num_batches()): # Placeholder function
    # Load batch data
    inputs, labels = get_training_batch(step) # Placeholder function

    # Train and predict
    train_model_on_batch(inputs, labels) # Placeholder function
    predictions = predict_on_batch(inputs) # Placeholder function

    # Add batch data to table
    for input_item, label, prediction in zip(inputs, labels, predictions):
        table.add_data(step, input_item, label, prediction)

    # Log the table incrementally
    wandb.log({"training_table": table}, step=step)

run.finish()
```

In the previous code example, the table is logged once per training step (`step`).

> Incremental logging is generally more computationally efficient than logging a new table each time (`log_mode=MUTABLE`). However, it is important to note that the W&B App may not render all rows in the table if you log a large number of increments. In this case, consider combining `INCREMENTAL` and `IMMUTABLE` logging modes to reduce the number of increments shown in the W&B App.

Reviewer: ```suggestion
Incremental logging is generally more computationally efficient than logging a new table each time (`log_mode=MUTABLE`). However, it is important to note that the W&B App may not render all rows in the table if you log a large number of increments. If your goal is to update and view your table data while your run is ongoing and to have all the data available for analysis, consider using both `INCREMENTAL` and `IMMUTABLE` logging modes.

```

---
title: Log tables
weight: 1
---

> Visualize and log tabular data with W&B Tables. A table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run]({{< relref "/guides/models/track/runs/" >}}). Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.

Reviewer 2: ```suggestion
Visualize and log tabular data with W&B Tables. A W&B Table is a two-dimensional grid of data where each column has a single type of data. Each row represents one or more data points logged to a W&B [run]({{< relref "/guides/models/track/runs/" >}}). Tables support primitive and numeric types, as well as nested lists, dictionaries, and rich media types.
```
Trying to disambiguate but I wonder if you should say that W&B Table and "table" are the same thing in this page.

Author: Proper vs common noun (The W&B Table product vs a table).

Reviewer 2: Right, but "A table is a two-dimensional grid..." seems like a tautology, that's the definition of a lowercase-T table in the world, not the W&B Tables-specific definition of a Table.

> Tables are a specialized [data type]({{< relref "/ref/python/data-types/" >}}) in W&B that are logged internally in W&B as [artifact]({{< relref "/guides/core/artifacts/" >}}) objects.

Reviewer 2: ```suggestion
A Table object is a specialized [data type]({{< relref "/ref/python/data-types/" >}}) that is logged internally in W&B as a type of [Artifact]({{< relref "/guides/core/artifacts/" >}}).
```
Maybe do a search for "tables" and see which would be better in the singular.

> You [create and log tables]({{< relref "#create-and-log-a-new-table" >}}) using the W&B Python SDK. When you create a table object, you specify the columns and data for the table and a [mode]({{< relref "#table-logging-modes" >}}). The mode determines how the table is logged and updated during your ML experiments.

Reviewer 2: ```suggestion
You [create and log Tables]({{< relref "#create-and-log-a-new-table" >}}) using the W&B Python SDK. When you create a Table, you specify its columns and data, along with a [mode]({{< relref "#table-logging-modes" >}}). The mode determines how the table is logged and updated by your ML experiments.
```
Again there are a few different ways we refer to this Table thing in a very short run.

Author: Let's stick with using lower case when referring to any table and upper case when referring to The product. Like how we do with artifacts vs W&B Artifacts. etc.

Reviewer 2: I worry that we are going to need to describe the Table panel, which may lead to prose like "A Table shows a table of....".

Author: We avoid using the product name in that case. E.g. We don't say "run a run..."

## Create and log a table

> 1. First, create a new run with `wandb.init()`. 

Reviewer 2: You don't need to number the prose in a numbered list.

3. Log the table to W&B with `run.log()`.

The following example shows how to create and log a table with two columns, `a` and `b`, and two rows of data, `["a1", "b1"]` and `["a2", "b2"]`:

```python
import wandb

# Start a new run
run = wandb.init(project="table-demo")

# Create a table object with two columns and two rows of data
my_table = wandb.Table(
    columns=["a", "b"],
    data=[["a1", "b1"], ["a2", "b2"]],
    log_mode="IMMUTABLE"
    )

# Log the table to W&B
run.log({"Table Name": my_table})

# Finish the run
run.finish()
```

## Logging modes

The `log_mode` parameter of the [`wandb.Table`]({{< relref "/ref/python/data-types/table" >}}) class determines how the table is logged and updated during your ML experiments. The `log_mode` parameter accepts three values: `IMMUTABLE`, `MUTABLE`, and `INCREMENTAL`. Each mode has different implications for how the table is logged, how it can be modified, and how it is rendered in the W&B App.

The modes are:

> - `IMMUTABLE`: Once a table is logged to W&B, you can not modify it.

Reviewer 2: ```suggestion
- `IMMUTABLE`: Once a table is logged to W&B, you cannot modify it.
```
Is this the default? Your example sets it manually but doesn't explain it in a comment, so I wasn't sure.

> - `MUTABLE`: After you log the table to W&B, you can overwrite the existing table by replacing that table with a new one.

Reviewer 2: ```suggestion
- `MUTABLE`: After you log the table to W&B, you can overwrite the entire existing table with a new one.
```

- `INCREMENTAL`: Add batches of new rows to a table throughout the machine learning experiment.

The following table describes the differences between the three modes and common use cases for each:

> | Mode  | Use Cases  | Benefits  |

Reviewer 2: The table could be combined with the prose explainations just above.

| ----- | ---------- | ----------|

> | `IMMUTABLE`   | - Storing tabular data generated at the end of a run for further analysis                              | - Minimal overhead when logged at the end of a run<br>- All rows rendered in UI |

Reviewer 2: It would be better to use a manual `<ul><li>....</li></ul>` within each cell. That helps you avoid the manual `<br>` and makes it more clear your intent than using the hyphens (which I don't think will get rendered as HTML list items).

| `MUTABLE`     | - Adding columns or rows to existing tables<br>- Enriching results with new information                        | - Capture Table mutations<br>- All rows rendered in UI                          |
| `INCREMENTAL` | - Adding rows to tables in batches<br> - Long-running training jobs<br>- Processing large datasets in batches<br>- Monitoring ongoing results | - View updates on UI during training<br>- Ability to step through increments    |

The next sections show example code snippets for each mode along with considerations for when to use each mode.

### MUTABLE mode

> `MUTABLE` mode updates an existing table by replacing the existing table with a new one. `MUTABLE` mode is useful when you want to add new columns and rows to an existing table in a non iterative process. If you want to add new batches of rows incrementally like in a training loop, consider using [`INCREMENTAL` mode]({{< relref "#INCREMENTAL-mode" >}}) instead.

Reviewer 2: Do we really need a summary, then a table, then a subsection for each? Also each of the three options is in a different order in each one of these.

Author: Condense. Added this list to the table.

| `MUTABLE`     | - Adding columns or rows to existing tables<br>- Enriching results with new information                        | - Capture Table mutations<br>- All rows rendered in UI                          |
| `INCREMENTAL` | - Adding rows to tables in batches<br> - Long-running training jobs<br>- Processing large datasets in batches<br>- Monitoring ongoing results | - View updates on UI during training<br>- Ability to step through increments    |

The next sections show example code snippets for each mode along with considerations for when to use each mode.

### MUTABLE mode

`MUTABLE` mode updates an existing table by replacing the existing table with a new one. `MUTABLE` mode is useful when you want to add new columns and rows to an existing table in a non iterative process. If you want to add new batches of rows incrementally like in a training loop, consider using [`INCREMENTAL` mode]({{< relref "#INCREMENTAL-mode" >}}) instead.

The following example shows how to create a table in `MUTABLE` mode, log it, and then add new columns to it.

{{% alert %}}

> The following example uses a placeholder function `load_eval_data()` to load data and a placeholder function `model.predict()` to make predictions. You will need to replace these with your own data loading and prediction functions.

Reviewer 2: It's a shame we can't inline the placeholder functions so that they get a sense of what it should do.

{{% /alert %}}

```python
import wandb
import numpy as np

run = wandb.init(project="mutable-table-demo")

# Create a table object with MUTABLE logging mode
table = wandb.Table(columns=["input", "label", "prediction"],
                    log_mode="MUTABLE")

# Load data and make predictions
inputs, labels = load_eval_data() # Placeholder function
raw_preds = model.predict(inputs) # Placeholder function

for inp, label, pred in zip(inputs, labels, raw_preds):
    table.add_data(inp, label, pred)

# Step 1: Log initial data 
wandb.log({"eval_table": table})  # Log initial table

# Step 2: Add confidence scores (e.g. max softmax)
confidences = np.max(raw_preds, axis=1)
table.add_column("confidence", confidences)
run.log({"eval_table": table})  # Add confidence info

# Step 3: Add post-processed predictions
# (e.g., thresholded or smoothed outputs)
post_preds = (confidences > 0.7).astype(int)
table.add_column("final_prediction", post_preds)
wandb.log({"eval_table": table})  # Final update with another column

run.finish()
```

> In the previous example, the table object is logged three times: once with the initial data, once with the confidence scores, and once with the final predictions.

Reviewer 2: ```suggestion
In the example, the table object is logged three times: once with the initial data, once with the confidence scores, and once with the final predictions.
```
Maybe move this up above the example.

{{% alert %}}

> Internally, the table object is replaced each time you log the table. Overwriting a table with a new one is computationally expensive and can be slow for large tables.

Reviewer 2: ```suggestion
In `MUTABLE` mode, the table object is replaced each time you log the table. Overwriting a table with a new one is computationally expensive and can be slow for large tables. Consider using `IMCREMENTAL` mode instead for large tables.
```
Is this what the alert is trying to get me to consider?

Author: Almost. You can't use INCREMENTAL mode for logging columns. You can only append new rows.

{{% /alert %}}

### INCREMENTAL mode

> In incremental mode, you log batches of rows to a table throughout the machine learning experiment. This

Reviewer 2: ```suggestion
In incremental mode, you log batches of rows to a table during the machine learning experiment. This
```
Maybe soft-wrap these lines instead of hard wrap.

is ideal for monitoring long-running jobs or when working with large tables that would be inefficient to log during the run for updates.

> You can not add new columns to a table in `INCREMENTAL` mode. If you need to add new columns to a table that you already logged, consider using `MUTABLE` mode instead.

Reviewer 2: ```suggestion
You cannot add new columns to a table in `INCREMENTAL` mode. To add new columns to a table that you already logged, consider using `MUTABLE` mode instead.
```
Maybe a scrub for "can not".

> The following example shows how to create a table in `INCREMENTAL` mode, log it, and then add new rows to it.

Reviewer 2: ```suggestion
The following example creates a table in `INCREMENTAL` mode, logs it, and then adds new rows to it.
```

{{% alert %}}

The following example uses a placeholder function `get_training_batch()` to load data, a placeholder function `train_model_on_batch()` to train the model, and a placeholder function `predict_on_batch()` to make predictions. You will need to replace these with your own data loading, training, and prediction functions.

> {{% /alert %}}

Reviewer 2: ```suggestion
{{% alert %}}
The following example uses a placeholder function `get_training_batch()` to load data, a placeholder function `train_model_on_batch()` to train the model, and a placeholder function `predict_on_batch()` to make predictions. You will need to replace these with your own data loading, training, and prediction functions.
{{% /alert %}}
```

```python
import wandb

run = wandb.init(project="incremental-table-demo")

# Create a table with INCREMENTAL logging mode
table = wandb.Table(columns=["step", "input", "label", "prediction"],
                    log_mode="INCREMENTAL")

# Training loop
for step in range(get_num_batches()): # Placeholder function
    # Load batch data
    inputs, labels = get_training_batch(step) # Placeholder function

    # Train and predict
    train_model_on_batch(inputs, labels) # Placeholder function
    predictions = predict_on_batch(inputs) # Placeholder function

    # Add batch data to table
    for input_item, label, prediction in zip(inputs, labels, predictions):
        table.add_data(step, input_item, label, prediction)

    # Log the table incrementally
    wandb.log({"training_table": table}, step=step)

run.finish()
```

> In the previous code example, the table is logged once per training step (`step`).

Reviewer 2: ```suggestion
In this example, the table is logged once per training `step`.
```
Maybe move this above the. code block.

> Incremental logging is generally more computationally efficient than logging a new table each time (`log_mode=MUTABLE`). However, it is important to note that the W&B App may not render all rows in the table if you log a large number of increments. If your goal is to update and view your table data while your run is ongoing and to have all the data available for analysis, consider using both `INCREMENTAL` and `IMMUTABLE` logging modes.

Reviewer 2: ```suggestion
Incremental logging is generally more computationally efficient than logging a new table each time (`log_mode=MUTABLE`). However, the W&B App may not render all rows in the table if you log a large number of increments. If your goal is to update and view your table data while your run is ongoing and to have all the data available for analysis, consider using both `INCREMENTAL` and `IMMUTABLE` logging modes.
```
This is the first time I'm reading that you can combine logging modes. Is it worth adding another heading for it?

Reviewer: To tag onto @mdlinville 's initial reaction, I think we can make it more explicit that we are using two tables each with a different log mode to avoid the users thinking that a Table's log mode can be changed mid run.

```suggestion
Incremental logging is generally more computationally efficient than logging a new table each time (`log_mode=MUTABLE`). However, the W&B App may not render all rows in the table if you log a large number of increments. If your goal is to update and view your table data while your run is ongoing and to have all the data available for analysis, consider using two tables. One with `INCREMENTAL` log mode and one with `IMMUTABLE` log mode.
```

{{% alert %}}

> Run workspaces in the W&B App have a limit of 100 increments. If you log more than 100 increments, only the most recent 100 are shown in the run workspace.

Reviewer 2: This seems a bit buried here. Could it be moved up a bit?

{{% /alert %}}

> As an example, consider the follow code snippet that combines `INCREMENTAL` and `IMMUTABLE` logging modes:

Reviewer 2: ```suggestion
This example combines `INCREMENTAL` and `IMMUTABLE` logging modes:
```

```python
import wandb

run = wandb.init(project="combined-logging-example")

# Create an incremental table for efficient updates during training
incr_table = wandb.Table(columns=["step", "input", "prediction", "label"],
                         log_mode="INCREMENTAL")

# Training loop
for step in range(get_num_batches()):
    # Process batch
    inputs, labels = get_training_batch(step)
    predictions = model.predict(inputs)

    # Add data to incremental table
    for inp, pred, label in zip(inputs, predictions, labels):
        incr_table.add_data(step, inp, pred, label)

    # Log the incremental update (suffix with -incr to distinguish from final table)
    run.log({"table-incr": incr_table}, step=step)

# At the end of training, create a complete immutable table with all data
# Using the default IMMUTABLE mode to preserve the complete dataset
final_table = wandb.Table(columns=incr_table.columns, data=incr_table.data, log_mode="IMMUTABLE")
run.log({"table": final_table})

run.finish()
```

> In the previous example, the `incr_table` is logged incrementally (with `log_mode="INCREMENTAL"`) during training. This allows you to log and view updates to the table as new data is processed. At the end of training, an immutable table (`final_table`) is created with all data from the incremental table. The immutable table is logged to preserve the complete dataset for further analysis and it enables you to view all rows in the W&B App. 

Reviewer 2: ```suggestion
In this example, the `incr_table` is logged incrementally (with `log_mode="INCREMENTAL"`) during training. This allows you to log and view updates to the table as new data is processed. At the end of training, an immutable table (`final_table`) is created with all data from the incremental table. The immutable table is logged to preserve the complete dataset for further analysis and it enables you to view all rows in the W&B App. 
```
Consider moving this up above the code block.

> ## Examples 

Reviewer 2: Maybe the previous sections were "Basic examples" and this is "More complex examples" or something like that?

Author: Ahh. Not a fan of using "basic", "advanced", " complex", etc.. What's "complex" to some might not be complex for others.

Reviewer 2: I guess what I mean is that we have just had a series of Examples and now we have more of them in a separate heading.

```python
import wandb
import numpy as np

run = wandb.init(project="mutable-logging")

# Step 1: Log initial predictions
table = wandb.Table(columns=["input", "label", "prediction"], log_mode="MUTABLE")
inputs, labels = load_eval_data()
raw_preds = model.predict(inputs)

for inp, label, pred in zip(inputs, labels, raw_preds):
    table.add_data(inp, label, pred)

run.log({"eval_table": table})  # Log raw predictions

# Step 2: Add confidence scores (e.g. max softmax)
confidences = np.max(raw_preds, axis=1)
table.add_column("confidence", confidences)
run.log({"eval_table": table})  # Add confidence info

# Step 3: Add post-processed predictions
# (e.g., thresholded or smoothed outputs)
post_preds = (confidences > 0.7).astype(int)
table.add_column("final_prediction", post_preds)
run.log({"eval_table": table})  # Final

run.finish()
```

### Resuming runs with INCREMENTAL tables

You can continue logging to an incremental table when resuming a run:

```python
# Start or resume a run
resumed_run = wandb.init(project="resume-incremental", id="your-run-id", resume="must")

# Create the incremental table; no need to populate with data from preivously logged table

> # Increments will be continue to be added to the Table artifact.

Reviewer: There was one final change that was made before merging in the feature. In order to resume adding increments to the table artifact, the user must not have summaries turned off. I think we should note this somewhere so there are no surprises.

Author: Nice catch @domphan-wandb.  How does a user check if summaries are on or off?

Reviewer: The user can disable summaries for a specific key via `wandb.define_metric("<table_key>", summary="none")` or all keys via `wandb.define_metric("*", summary="none")`

I think we can state something along the lines of:

If summaries for the table key is disabled, then the increments will be logged as a new table instead of continuing from the previously logged table (before the resume)

Author: Sanity check:
1. This only applies to custom metrics (i.e. `wandb.define_metric`)? 
2. The full code snippet would look like:  
```python
# Start or resume a run
resumed_run = wandb.init(project="resume-incremental", id="your-run-id", resume="must")

# Disable summaries for the table key
wandb.define_metric("loss", summary="none")  # Disable summaries for the table key

# Create the incremental table; no need to populate with data from preivously logged table
# Increments will be continue to be added to the Table artifact.
table = wandb.Table(columns=["step", "metric"], log_mode="INCREMENTAL")

# Continue logging
for step in range(resume_step, final_step):
    metric = compute_metric(step)
    table.add_data(step, metric)
    resumed_run.log({"loss": table}, step=step)

resumed_run.finish()
```
